### PR TITLE
7366 fix area occ estm

### PR DESCRIPTION
**Pull request by anyfilatov**

### CODE REVIEW

File: src/tiny_slam/tiny_grid_cells.h [hunk 1]
```diff
@@ -39,7 +39,9 @@ class BaseTinyCell : public GridCell {
    * \param[in] quality - the quality of the experiment value \f$\alpha\f$.
    */
   void set_value(const Occupancy &value, double quality) override {
-    _prob = (1.0 - quality) * _prob + quality * value.prob_occ;
+    if(!value.isNan()){
```
Comment (anyfilatov) [inline src/tiny_slam/tiny_grid_cells.h]: probably better change the name of function to "is_valid"

File: src/core/geometry_utils.h [hunk 2]
```diff
@@ -11,8 +11,35 @@
 #include <cmath>
 
 
-#define EQ_DOUBLE(a, b)                         \
-  (std::abs((a) - (b)) < 0.0001)
+inline bool equal(double a, double b) {
+  return (std::abs((a) - (b)) < 0.0001);
```
Comment (HatlessFox) [inline src/core/geometry_utils.h]: Code improvement: replace 0.0001 with numeric_limits<>::epsilon()

File: src/core/geometry_utils.h [hunk 3]
```diff
@@ -11,8 +11,35 @@
...
+inline bool equal(double a, double b, double c) {
+  return (equal(a, b) && equal(a, c));
+}
+
+inline bool less_or_equal(double a, double b) {
+  return (equal(a, b) || (a < b));
+}
+
+inline bool is_between(double point, double left, double right) {
+  return (less_or_equal(left, point) && less_or_equal(point, right)) ||
```
Comment (HatlessFox) [inline src/core/geometry_utils.h]: 1. If we need to check both cases ((l < r) and (r < l)) the arg names seem confusing to me (the "left" is supposed to be less than the "right").
2. If you choose to rename args please align the OR clauses.

File: src/core/geometry_utils.h [hunk 4]
```diff
@@ -11,8 +11,35 @@
...
+      (less_or_equal(right, point) && less_or_equal(point, left));
+}
+
+inline bool is_strictly_between(double point, double left, double right) {
```
Comment (HatlessFox) [inline src/core/geometry_utils.h]: Similar to is_between comment.
Swap the point and the left arguments to make argument order more intuitive.

File: src/core/geometry_utils.h [hunk 5]
```diff
@@ -11,8 +11,35 @@
...
+  return (((left < point) && (point < right)) ||
+        ((right < point) && (point < left)));
+}
+
+inline bool both_less(double x, double y, double threshold) {
```
Comment (HatlessFox) [inline src/core/geometry_utils.h:38]: move the threshold argument to the first position.

File: src/core/geometry_utils.h [hunk 6]
```diff
@@ -38,10 +65,21 @@ struct Rectangle {
    * \param x,y Coordinates of a target point.
    * \return True if the rectangle contains a point and False otherwise.
    */
-  bool does_contain(double x, double y) const {
-    return ((bot < y) && (y < top)) && ((left < x) && (x < right));
+  bool contains(double x, double y) const {
+    return (is_between(y, bot, top) && is_between(x, left, right));
+  }
+
+  bool contains_on_border(double x, double y) const {
```
Comment (HatlessFox) [inline src/core/geometry_utils.h]: Better: is_on_border

File: src/core/geometry_utils.h [hunk 7]
```diff
@@ -38,10 +65,21 @@ struct Rectangle {
...
+    return ((equal(x, left) && is_between(y, bot, top)) ||
+            (equal(x, right) && is_between(y, bot, top)) ||
+            (equal(y, top) && is_between(x, left, right)) ||
+            (equal(y, bot) && is_between(x, left, right)));
   }
 
+  bool contains_in_corner(double x, double y) const {
```
Comment (HatlessFox) [inline src/core/geometry_utils.h]: Better: is_in_corner

File: src/core/geometry_utils.h [hunk 8]
```diff
@@ -38,10 +65,21 @@ struct Rectangle {
    * \param x,y Coordinates of a target point.
    * \return True if the rectangle contains a point and False otherwise.
    */
-  bool does_contain(double x, double y) const {
-    return ((bot < y) && (y < top)) && ((left < x) && (x < right));
+  bool contains(double x, double y) const {
```
Comment (HatlessFox) [inline src/core/geometry_utils.h]: Consider: is_inside

File: src/core/geometry_utils.h [hunk 9]
```diff
@@ -50,6 +88,17 @@ struct Rectangle {
     return (top - bot)*(right - left);
   }
 
+  double get_height() const {
```
Comment (HatlessFox) [inline src/core/geometry_utils.h]: drop the "get_" part

File: src/core/maps/area_occupancy_estimator.h [hunk 10]
```diff
@@ -81,9 +81,68 @@ class AreaOccupancyEstimator : public CellOccupancyEstimator {
   virtual Occupancy estimate_occupancy(const Beam &beam,
                                        const Rectangle &cell_bnds,
                                        bool is_occ) override {
-    Intersections intrs = find_intersections(beam, cell_bnds, is_occ);
-    double chunk_area = compute_chunk_area(beam, cell_bnds, is_occ, intrs);
-    return estimate_occupancy(chunk_area, cell_bnds.area(), is_occ);
+    Occupancy result(0.0, 0.0);
+    if (equal (beam.x_st, beam.x_end) && (equal(beam.y_st, beam.y_end))) {
```
Comment (HatlessFox) [inline src/core/maps/area_occupancy_estimator.h]: Space after the first equal is redundant.
Move the condition computation to the separate variable with a meaningful name.

File: src/core/maps/area_occupancy_estimator.h [hunk 11]
```diff
@@ -81,9 +81,68 @@ class AreaOccupancyEstimator : public CellOccupancyEstimator {
...
+      return result;
+    }
+    if(beam.encounters_rect(cell_bnds)){
+      return is_occ ? Occupancy(1.0, 1.0) : Occupancy(base_empty_prob(), 0.5);
+    }
+    if(!cell_bnds.contains(beam.x_end, beam.y_end) && is_occ) {
+      result.setInvalid();
+      return result;
+    }
+    Beam buf_beam = beam.generate_revert();
```
Comment (HatlessFox) [inline src/core/maps/area_occupancy_estimator.h]: What does the "buf_" prefix mean?

File: src/core/maps/area_occupancy_estimator.h (hunk 10 above)
Comment (HatlessFox) [inline src/core/maps/area_occupancy_estimator.h]: Can we merge 1st, 2nd, 4th "ifs" into one?

File: src/core/maps/area_occupancy_estimator.h [hunk 12]
```diff
@@ -81,9 +81,68 @@ class AreaOccupancyEstimator : public CellOccupancyEstimator {
...
+    if(buf_beam.encounters_rect(cell_bnds)) {
+      result.setInvalid();
+      return result;
+    }
+
+    Beam n_beam = beam;
+    bool flag_changed_start_end = false;
+    if(!cell_bnds.contains_on_border(beam.x_st, beam.y_st) &&
```
Comment (HatlessFox) [inline src/core/maps/area_occupancy_estimator.h]: Can we move this code into separate function?

File: src/core/maps/area_occupancy_estimator.h [hunk 13]
```diff
@@ -81,9 +81,68 @@ class AreaOccupancyEstimator : public CellOccupancyEstimator {
...
+      result.setInvalid();
+      return result;
+    }
+    Beam buf_beam = beam.generate_revert();
+    if(buf_beam.encounters_rect(cell_bnds)) {
+      result.setInvalid();
+      return result;
+    }
+
+    Beam n_beam = beam;
```
Comment (HatlessFox) [inline src/core/maps/area_occupancy_estimator.h]: What does the "n_" prefix mean?

File: src/core/maps/area_occupancy_estimator.h [hunk 14]
```diff
@@ -81,9 +81,68 @@ class AreaOccupancyEstimator : public CellOccupancyEstimator {
...
+        if (cell_bnds.contains(beam.x_st, beam.y_st)) {
+          n_beam = beam.generate_revert();
+          flag_changed_start_end = true;
+        }
+      }
+    }
+
+    Intersections intrs = find_intersections(n_beam, cell_bnds, is_occ);
+    double chunk_area = compute_chunk_area(n_beam, cell_bnds, is_occ, intrs);
+    double chunk_proportion = chunk_area/cell_bnds.area();
```
Comment (HatlessFox) [inline src/core/maps/area_occupancy_estimator.h]: Add spaces around "/"

Comment (HatlessFox) [inline src/core/maps/area_occupancy_estimator.h]: Consider the "area_ratio" name which is shorter

File: src/core/maps/area_occupancy_estimator.h [hunk 15]
```diff
@@ -81,9 +81,68 @@ class AreaOccupancyEstimator : public CellOccupancyEstimator {
...
+    result = estimate_occupancy(chunk_area, cell_bnds.area(), is_occ);
+
+    chunk_proportion = (0.5 < chunk_proportion) ?
+                         (1 - chunk_proportion) : chunk_proportion;
+    if(flag_changed_start_end) {
```
Comment (HatlessFox) [inline src/core/maps/area_occupancy_estimator.h]: This "if" is not intuitive. Change the cond var name to something more meaningful.

File: src/core/maps/area_occupancy_estimator.h [hunk 16]
```diff
@@ -81,9 +81,68 @@ class AreaOccupancyEstimator : public CellOccupancyEstimator {
...
+      result.prob_occ = 0.01;
+    }
+    if(n_beam.tangents_rect(cell_bnds)) {
+      result.estimation_quality = 0.01;
+    }
+    else if(n_beam.reaches_border_passing_through_rect(cell_bnds)) {
+      result.prob_occ = base_empty_prob();
+      result.estimation_quality = chunk_proportion;
+    }
+    if (equal(result.prob_occ, base_empty_prob()) && is_occ) {
```
Comment (HatlessFox) [inline src/core/maps/area_occupancy_estimator.h:72]: It would be better to somehow move the check closer to the method's beginning.

File: src/core/maps/cell_occupancy_estimator.h [hunk 17]
```diff
@@ -37,8 +37,17 @@ struct Occupancy {
     prob_occ(prob), estimation_quality(quality), x(curr_x), y(curr_y) {}
 
   bool operator==(const Occupancy &that) {
-    return EQ_DOUBLE(prob_occ, that.prob_occ) &&
-           EQ_DOUBLE(estimation_quality, that.estimation_quality);
+    return equal(prob_occ, that.prob_occ) &&
+           equal(estimation_quality, that.estimation_quality);
+  }
+
+  void setInvalid(){
```
Comment (HatlessFox) [inline src/core/maps/cell_occupancy_estimator.h]: better: invalidade.

File: src/core/maps/cell_occupancy_estimator.h [hunk 18]
```diff
@@ -37,8 +37,17 @@ struct Occupancy {
...
+    prob_occ = std::numeric_limits<double>::quiet_NaN();
+    estimation_quality = std::numeric_limits<double>::quiet_NaN();
+  }
+
+  bool isValid() const {
```
Comment (HatlessFox) [inline src/core/maps/cell_occupancy_estimator.h]: Method name should be in snake case.

File: src/core/maps/cell_occupancy_estimator.h [hunk 19]
```diff
@@ -47,6 +56,80 @@ struct Occupancy {
 struct Beam {
   double x_st, y_st; ///< Coordinates of a start of the beam.
   double x_end, y_end; ///< Coordinates where the beam reaches an obstacle.
+
+  //methods
+  bool tangents_rect(const Rectangle& bnds) const {
```
Comment (HatlessFox) [inline src/core/maps/cell_occupancy_estimator.h]: better: touches. AFAIK tangent is not a verb.

File: src/core/maps/cell_occupancy_estimator.h [hunk 20]
```diff
@@ -47,6 +56,80 @@ struct Occupancy {
...
+    return equal(x_st, x_end, bnds.left) || equal(y_st, y_end, bnds.bot) ||
+           equal(x_st, x_end, bnds.right) || equal(y_st, y_end, bnds.top);
+  }
+
+  bool intersects_rect(const Rectangle& bnds) const {
```
Comment (HatlessFox) [inline src/core/maps/cell_occupancy_estimator.h]: Drop the "rest" suffix

File: src/core/maps/cell_occupancy_estimator.h [hunk 21]
```diff
@@ -47,6 +56,80 @@ struct Occupancy {
...
+    bool result = false;
+    if(bnds.contains(x_st, y_st) || bnds.contains(x_end, y_end)) {
```
Comment (HatlessFox) [inline src/core/maps/cell_occupancy_estimator.h]: Add a space after "if", check other ifs.

File: src/core/geometry_utils.h [hunk 22]
```diff
@@ -11,8 +11,35 @@
...
+inline bool is_strictly_between(double point, double left, double right) {
+  return (((left < point) && (point < right)) ||
+        ((right < point) && (point < left)));
+}
+
+inline bool both_less(double x, double y, double threshold) {
+  return (x < threshold) && (y < threshold);
+}
+
+inline bool both_higher(double x, double y, double threshold) {
```
Comment (HatlessFox) [inline src/core/geometry_utils.h]: s/high/great

File: src/core/maps/cell_occupancy_estimator.h [hunk 23]
```diff
@@ -47,6 +56,80 @@ struct Occupancy {
...
+    if (equal(y_end, bnds.bot) || equal(y_end, bnds.top)) {
+      return is_strictly_between(y_end, y_st, bnds.center_y());
+    }
+    return false;
+  }
+
+  bool reaches_border_passing_through_rect(const Rectangle& bnds) const {
+    return bnds.contains_on_border(x_end, y_end) && intersects_rect(bnds);
+  }
+  Beam generate_revert() const {
```
Comment (HatlessFox) [inline src/core/maps/cell_occupancy_estimator.h]: rename: revert

File: src/tiny_slam/tiny_grid_cells.h [hunk 24]
```diff
@@ -39,7 +39,9 @@ class BaseTinyCell : public GridCell {
    * \param[in] quality - the quality of the experiment value \f$\alpha\f$.
    */
   void set_value(const Occupancy &value, double quality) override {
-    _prob = (1.0 - quality) * _prob + quality * value.prob_occ;
```
Comment (HatlessFox) [inline src/tiny_slam/tiny_grid_cells.h:42]: I prefer:
if (not valid)

> return

formula

File: src/core/maps/cell_occupancy_estimator.h [hunk 25]
```diff
@@ -47,6 +56,80 @@ struct Occupancy {
...
+             is_strictly_between(x_st, bnds.left, bnds.right));
+    }
+
+    if(both_less(x_st, x_end, bnds.left) ||
+       both_higher(x_st, x_end, bnds.right) ||
+       both_less(y_st, y_end, bnds.bot) ||
+       both_higher(y_st, y_end, bnds.top)) {
+         return false;
+    }
+    double k = (y_end - y_st) / (x_end - x_st);
```
Comment (HatlessFox) [inline src/core/maps/cell_occupancy_estimator.h]: Well, we've already implemented the code that finds intersections (find_intersections) method. Does the code below makes the same?

File: src/core/geometry_utils.h (hunk 4 above)
Comment (anyfilatov) [reply src/core/geometry_utils.h]: if the point and the left arguments are swapped, shouldn't the name of function be changed to smth like "are_ordered"?

File: src/core/geometry_utils.h (hunk 5 above)
Comment (anyfilatov) [reply src/core/geometry_utils.h:38]: Will it be helpful for understanding?
Shouldn't the parameters be placed in order from lower to greater?

File: src/core/maps/area_occupancy_estimator.h (hunk 11 above)
Comment (anyfilatov) [reply src/core/maps/area_occupancy_estimator.h]: I needed a variable for next four lines of code. I decided to call it "buffer"

File: src/core/maps/area_occupancy_estimator.h (hunk 13 above)
Comment (anyfilatov) [reply src/core/maps/area_occupancy_estimator.h]: "n_" for "new". In case variable beam cannot be changed, I have to create another variable, that would be used instead

File: src/core/maps/cell_occupancy_estimator.h (hunk 25 above)
Comment (anyfilatov) [reply src/core/maps/cell_occupancy_estimator.h]: The method, you've implemented outputs a vector of intersections. For the purpose of this function it is redundant

File: src/core/maps/cell_occupancy_estimator.h (hunk 23 above)
Comment (anyfilatov) [reply src/core/maps/cell_occupancy_estimator.h]: I think that the name "revert" means that the beam is going to be changed, but this function creates another beam

File: src/tiny_slam/tiny_grid_cells.h [hunk 26]
```diff
@@ -39,7 +39,9 @@ class BaseTinyCell : public GridCell {
...
+    if(value.isValid()){
+      _prob = (1.0 - quality) * _prob + quality * value.prob_occ;
+    }
   }
 private:
   double _prob;
```
Comment (anyfilatov) [inline src/tiny_slam/tiny_grid_cells.h:48]: with all other comments I agree - I will correct my code

File: src/core/geometry_utils.h (hunk 5 above)
Comment (HatlessFox) [reply src/core/geometry_utils.h:38]: Yes, they should, my bad. The comment is for the both_higher function.

File: src/core/geometry_utils.h (hunk 4 above)
Comment (HatlessFox) [reply src/core/geometry_utils.h]: It's up to you. Both options are ok to me. In case you pick the second one use the "are_strictly_ordered" name.

File: src/core/maps/area_occupancy_estimator.h (hunk 11 above)
Comment (HatlessFox) [reply src/core/maps/area_occupancy_estimator.h]: But it is definitely not a buffer. Why can't you just write

beam.revert().encounters_rect(...)

File: src/core/maps/area_occupancy_estimator.h (hunk 13 above)
Comment (HatlessFox) [reply src/core/maps/area_occupancy_estimator.h]: The "new_beam" name isn't meaningful in this context, is it? You need to come up with a better name.
I suggest to think about refactoring that allows to drop the variable.

File: src/core/maps/cell_occupancy_estimator.h (hunk 25 above)
Comment (HatlessFox) [reply src/core/maps/cell_occupancy_estimator.h]: My main objection is the DRY principle violation.
1) What do you mean by redundant? Can you use output of the method to test whether a ray intersects a cell? I bet, you can (especially if you try to find a case when the cell isn't intersected)
2) Have tried to fix the compute_chunk_area method in order to handle the corner cases when a ray touches cell's side and when a ray doesn't intersect with a cell.

File: src/core/maps/cell_occupancy_estimator.h (hunk 23 above)
Comment (HatlessFox) [reply src/core/maps/cell_occupancy_estimator.h]: I don't think so. Consider operator-() (unary minus). An object modification is up to the operator implementation. The const tells a class user that а retuned value is a new one. I believe the same logic is relevant to our case.

If you want to stick to the "generate_revert" name then create a static method in the Beam class that performs reversion and returns a new value.

By the way, why can't me move the method to the unary minus operator?

File: src/core/maps/cell_occupancy_estimator.h [hunk 27]
```diff
@@ -47,6 +56,73 @@ struct Occupancy {
...
   double x_end, y_end; ///< Coordinates where the beam reaches an obstacle.
+
+  //methods
+  bool touches_rect(const Rectangle& bnds) const {
+    return equal(x_st, x_end, bnds.left) || equal(y_st, y_end, bnds.bot) ||
+           equal(x_st, x_end, bnds.right) || equal(y_st, y_end, bnds.top);
+  }
+
+  Beam operator- () const {
+    return (*this).revert();
```
Comment (HFX-TA) [inline src/core/maps/cell_occupancy_estimator.h]: Can we move the revert implementation inside the operator?

File: src/core/maps/cell_occupancy_estimator.h [hunk 28]
```diff
@@ -47,6 +56,73 @@ struct Occupancy {
...
+  }
+
+
```
Comment (HatlessFox) [inline src/core/maps/cell_occupancy_estimator.h]: Redundant empty lines

File: src/core/maps/area_occupancy_estimator.h [hunk 29]
```diff
@@ -81,31 +32,62 @@ class AreaOccupancyEstimator : public CellOccupancyEstimator {
...
+        beam_escapes_occ_cell || beam_escapes_from_border) {
+      result.invalidate();
+      return result;
+    }
+
+    if (beam.encounters(cell_bnds)){
+      return is_occ ? Occupancy(1.0, 1.0) : Occupancy(base_empty_prob(), 0.5);
+    }
+
+    Beam remote_beam = beam.move_start_out(cell_bnds);
```
Comment (HatlessFox) [inline src/core/maps/area_occupancy_estimator.h]: I don't get why the beam is "remote". Mb acting_beam is a better name?

File: src/core/maps/cell_occupancy_estimator.h [hunk 30]
```diff
@@ -47,6 +56,73 @@ struct Occupancy {
...
+    if (equal(x_end, bnds.left) || equal(x_end, bnds.right)) {
+      return are_strictly_ordered(x_st, x_end, bnds.center_x());
+    }
+    if (equal(y_end, bnds.bot) || equal(y_end, bnds.top)) {
+      return are_strictly_ordered(y_st, y_end, bnds.center_y());
+    }
+    return false;
+  }
+
+  bool reaches_border_passing_through(const Rectangle& bnds) const {
```
Comment (HatlessFox) [inline src/core/maps/cell_occupancy_estimator.h]: I suggest "stops_at_border_passing_through"

File: src/core/maps/cell_occupancy_estimator.h [hunk 31]
```diff
@@ -47,6 +56,73 @@ struct Occupancy {
 struct Beam {
   double x_st, y_st; ///< Coordinates of a start of the beam.
   double x_end, y_end; ///< Coordinates where the beam reaches an obstacle.
+
+  //methods
+  bool touches_rect(const Rectangle& bnds) const {
```
Comment (HatlessFox) [inline src/core/maps/cell_occupancy_estimator.h]: Mb "passes_through_border" оr something like that?

File: src/core/maps/cell_occupancy_estimator.h [hunk 32]
```diff
@@ -47,6 +56,73 @@ struct Occupancy {
...
+    bool result = false;
+    Ray ray(x_st, y_st, x_end - x_st, y_end - y_st);
+    Intersections intrs = ray.find_intersections(bnds);
+    for (auto intersection : intrs) {
+      result |= bnds.is_on_border(intersection.x, intersection.y);
+    }
+    return result;
+  }
+
+  bool encounters(const Rectangle& bnds) const {
```
Comment (HatlessFox) [inline src/core/maps/cell_occupancy_estimator.h]: This method tests whether the beam touches a given rect, am I right?

File: src/core/maps/cell_occupancy_estimator.h [hunk 33]
```diff
@@ -47,6 +56,73 @@ struct Occupancy {
...
+    return bnds.is_on_border(x_end, y_end) && intersects(bnds);
+  }
+  Beam revert() const {
+    Beam result(*this);
+    std::swap(result.x_st, result.x_end);
+    std::swap(result.y_st, result.y_end);
+    return result;
+  }
+
+  Beam move_start_out(const Rectangle& bnds) const {
```
Comment (HatlessFox) [inline src/core/maps/cell_occupancy_estimator.h]: "move_beam_start_out" sounds clearer to me. Taking into account a new beam creation, I recommend the "clone_with_start_outside" name.

File: src/core/maps/cell_occupancy_estimator.h (hunk 31 above)
Comment (anyfilatov) [reply src/core/maps/cell_occupancy_estimator.h]: passes through means that the beginning end ending are divided by a border. 
But this function checks whether a beam is passing through a cell, but doesn't contain any point inside cell except of points on border

File: src/core/maps/cell_occupancy_estimator.h (hunk 32 above)
Comment (anyfilatov) [reply src/core/maps/cell_occupancy_estimator.h]: this method checks whether a beam stops on a border of a cell and not passes through.
may be there is some misunderstanding: touches_rect means that the beam contains one border (or a piece of border). Encounters means that the beam stops on border, but no one point of beam contains inside a cell

File: src/core/maps/cell_occupancy_estimator.h (hunk 31 above)
Comment (HatlessFox) [reply src/core/maps/cell_occupancy_estimator.h]: Agreed.

File: src/core/maps/cell_occupancy_estimator.h (hunk 32 above)
Comment (HatlessFox) [reply src/core/maps/cell_occupancy_estimator.h]: Got it. As for me, "touch" and "encounter" are not the best verbs to express difference between a case when a beam contains a border and a case when a beam stops on a border.

Anton, I can't suggest more accurate verbs right now; what do you think about substitutions "touch" -> "contains_rect_border", "encounter" -> "stops_on_rect_border"?

Comment (anyfilatov) [reply src/core/maps/cell_occupancy_estimator.h]: I don't like the name "stops_on_border", because it could mean that beam passes through a rectangle and stops at the far bound.
As for me the verb "to encounter" is the most preferable. Moreover, in a code there are lines such as
"beam.encounters(cell_bnds)" and I guess that it is clear, that the beam reaches the bounds of cell and stops at the closest bound.

Comment (HatlessFox) [reply src/core/maps/cell_occupancy_estimator.h]: "Encounter" means "meet _unexpectedly_". I believe the name should emphasize both the fact of stoppage and a location of stoppage  (the front border). "stop_on/at_border" isn't good enough as well.

May be the team could come up with a better name. Should we ask them? We can also ask them about clearness of the "beam.encounter(cell_bnds)" statement.

Comment (anyfilatov) [reply src/core/maps/cell_occupancy_estimator.h]: My brother came up with "reaches_border". I think that it is a pretty good name

Comment (HatlessFox) [reply src/core/maps/cell_occupancy_estimator.h]: Agreed
